doc/man/man8/slapschema.8
.TH SLAPSCHEMA 8C "RELEASEDATE" "OpenLDAP LDVERSION"
.\" Copyright 1998-2011 The OpenLDAP Foundation All Rights Reserved.
.\" Copying restrictions apply.  See COPYRIGHT/LICENSE.
.\" $OpenLDAP: pkg/ldap/doc/man/man8/slapschema.8,v 1.1.2.5 2010/04/14 22:59:08 quanah Exp $
.SH NAME
slapschema \- SLAPD in-database schema checking utility
.SH SYNOPSIS
.B SBINDIR/slapschema
[\c
.BI \-a filter\fR]
[\c
.BI \-b suffix\fR]
[\c
.BR \-c ]
[\c
.BI \-d debug-level\fR]
[\c
.BI \-f slapd.conf\fR]
[\c
.BI \-F confdir\fR]
[\c
.BR \-g ]
[\c
.BI \-H URI\fR]
[\c
.BI \-l error-file\fR]
[\c
.BI \-n dbnum\fR]
[\c
.BI \-o option\fR[ = value\FR]]
[\c
.BI \-s subtree-dn\fR]
[\c
.BR \-v ]
.LP
.SH DESCRIPTION
.LP
.B Slapschema
is used to check schema compliance of the contents of a
.BR slapd (8)
database.
It opens the given database determined by the database number or
suffix and checks the compliance of its contents with the corresponding
schema. Errors are written to standard output or the specified file.
Databases configured as
.B subordinate
of this one are also output, unless \fB\-g\fP is specified.
.LP
Administrators may need to modify existing schema items, including
adding new required attributes to objectClasses,
removing existing required or allowed attributes from objectClasses,
entirely removing objectClasses,
or any other change that may result in making perfectly valid entries
no longer compliant with the modified schema.
The execution of the
.B slapschema tool after modifying the schema can point out
inconsistencies that would otherwise surface only when
inconsistent entries need to be modified.

.LP
The entry records are checked in database order, not superior first
order.  The entry records will be checked considering all
(user and operational) attributes stored in the database.
Dynamically generated attributes (such as subschemaSubentry)
will not be considered.
.SH OPTIONS
.TP
.BI \-a \ filter
Only check entries matching the asserted filter.
For example

slapschema \-a \\
    "(!(entryDN:dnSubtreeMatch:=ou=People,dc=example,dc=com))"

will check all but the "ou=People,dc=example,dc=com" subtree
of the "dc=example,dc=com" database.
Deprecated; use \fB-H\fP \fIldap:///???(filter)\fP instead.
.TP
.BI \-b \ suffix 
Use the specified \fIsuffix\fR to determine which database to
check.  The \fB\-b\fP cannot be used in conjunction
with the
.B \-n
option.
.TP
.B \-c
Enable continue (ignore errors) mode.
.TP
.BI \-d \ debug-level
Enable debugging messages as defined by the specified
.IR debug-level ;
see
.BR slapd (8)
for details.
.TP
.BI \-f \ slapd.conf
Specify an alternative
.BR slapd.conf (5)
file.
.TP
.BI \-F \ confdir
specify a config directory.
If both
.B \-f
and
.B \-F
are specified, the config file will be read and converted to
config directory format and written to the specified directory.
If neither option is specified, an attempt to read the
default config directory will be made before trying to use the default
config file. If a valid config directory exists then the
default config file is ignored.
.TP
.B \-g
disable subordinate gluing.  Only the specified database will be
processed, and not its glued subordinates (if any).
.TP
.B \-H \ URI
use dn, scope and filter from URI to only handle matching entries.
.TP
.BI \-l \ error-file
Write errors to specified file instead of standard output.
.TP
.BI \-n \ dbnum
Check the \fIdbnum\fR\-th database listed in the
configuration file. The config database
.BR slapd\-config (5),
is always the first database, so use
.B \-n 0

The
.B \-n
cannot be used in conjunction with the
.B \-b
option.
.TP
.BI \-o \ option\fR[ = value\fR]
Specify an
.I option
with a(n optional)
.IR value .
Possible generic options/values are:
.LP
.nf
              syslog=<subsystems>  (see `\-s' in slapd(8))
              syslog\-level=<level> (see `\-S' in slapd(8))
              syslog\-user=<user>   (see `\-l' in slapd(8))

.fi
.TP
.BI \-s \ subtree-dn
Only check entries in the subtree specified by this DN.
Implies \fB\-b\fP \fIsubtree-dn\fP if no
.B \-b
nor
.B \-n
option is given.
Deprecated; use \fB-H\fP \fIldap:///subtree-dn\fP instead.
.TP
.B \-v
Enable verbose mode.
.SH LIMITATIONS
For some backend types, your
.BR slapd (8)
should not be running (at least, not in read-write
mode) when you do this to ensure consistency of the database. It is
always safe to run 
.B slapschema
with the
.BR slapd\-bdb (5),
.BR slapd\-hdb (5),
and
.BR slapd\-null (5)
backends.
.SH EXAMPLES
To check the schema compliance of your SLAPD database after modifications
to the schema, and put any error in a file called
.BR errors.ldif ,
give the command:
.LP
.nf
.ft tt
	SBINDIR/slapcat \-l errors.ldif
.ft
.fi
.SH "SEE ALSO"
.BR ldap (3),
.BR ldif (5),
.BR slapd (8)
.LP
"OpenLDAP Administrator's Guide" (http://www.OpenLDAP.org/doc/admin/)
.SH ACKNOWLEDGEMENTS
.so ../Project


CHANGES
OpenLDAP 2.0 Change Log

OpenLDAP 2.0.27 Release
	Build environment
		Fix getopt(3) detection

OpenLDAP 2.0.26 Release
	Fixed slapd ACL buffer size bug (ITS#1963)
	Fixed slapd ACL filters w/substrings bug (ITS#2023, ITS#2027)
	Fxied slapd directory string space bug (ITS#2072)
	Fixed slurpd reject file handling bugs (ITS#2007, ITS#2009)
	Fixed back-ldbm indirect IDL bug (ITS#2039)
	Fixed libldap buffer size bugs (ITS#1935, ITS#1936)
	Fixed libldap gethostbyname mutex init bug (ITS#2019)
	Fixed liblutil passwd bug (ITS#2055)
	Fixed sasl large reads and writes (ITS#1983, also ITS#910,ITS#1861)
	Updated back-ldbm reindexing - slapindex rebuilds dn2id index
	Build environment
		Require version 1.5 if Cyrus SASL is present
		Remove tempnam checking and replacment implementation

OpenLDAP 2.0.25 Release
	Fixed slapd extended op reqdata crash bug (ITS#1721)

OpenLDAP 2.0.24 Release
	Fixed slapd max incoming macro bug (ITS#1828)
	Fixed slapd acl group/dnaddr bug (ITS#1607)
	Fixed slapd connection management bug (ITS#1655)
	Fixed slapd port ntoh / addr init bugs (ITS#1615, ITS#1728)
	Fixed slapd PF_lOCAL typo (ITS#1660)
	Fixed slapd extended ops with NULL reqdata bug (ITS#1721)
	Fixed slapd replog logging without replica (ITS#1335)
	Fixed slapd NULL attrval indexing (ITS#1843)
	Fixed slurpd tls init bug (ITS#1613)
	Fixed back-ldbm dn normalization bug in onelevel searches (ITS#1654)
	Fixed back-ldbm modrdn root dn check (ITS#1761)
	Fixed back-ldbm idl overrun bug (ITS#1570)
	Fixed back-ldbm passwd textlen bug (ITS#1742)
	Fixed ber_bvstrdup() empty string bug (ITS#1662)
	Fixed ldapsearch filter bug (ITS#1649)
	Fixed maildap NULL domain bug (ITS#1022)
	Fixed libldap hostless URL bug (ITS#1697)
	Fixed liblutil /dev/urandom incomplete reads (ITS#1834)
	Fixed libldap_r threadpool free list (ITS#1839)
	Fixed libldap_r NT threads cond_broadcast (ITS#1865)
	Fixed sasl writes > 64K (ITS#1861)
	Fixed sasl session callback corruption (ITS#1799)
	Build environment
		Removed lint (ITS#1744)
	Documentation
		Updated release documents

OpenLDAP 2.0.23 Release
	Fixed slapd moddn acl bug (ITS#1562) (ITS#1582)
	Fixed slapd acl regex (to dn="") portability bug (ITS#1579)
	Disabled -lldap atexit() handler (ITS#1577)
	Build environment
		Updated config.guess/sub scripts (ITS#1567)
	Documentation
		Fixed a number of typos (ITS#1578)

OpenLDAP 2.0.22 Release
	Fixed slapd passwd modify referral bug
	Fixed back-ldbm index threading bug
	Fixed back-ldbm ordering presence index bug
	Fixed back-ldap modify bug (ITS#1547)
	Fixed -lldap TLS errno bug
	Fixed ud v2 kbind bug
	Updated slapd slurpd friendly code
	Build environment
		Fixed Mach cthreads detection (ITS#1554)
		Updated Berkeley DB 4 detection

OpenLDAP 2.0.21 Release
	Fix slapd empty string indexing/filtering handling (ITS#1507)
	Changed slapd ACL selfwrite replace logic (ITS#1530)
	Build environment
		Updated IPv6 configuration argument handling
		Fix back-shell tools make depend bug (ITS#1518)
		Removed extraneous files

OpenLDAP 2.0.20 Release
	Fixed slapd ACL empty replace bug (ITS#1530)
	Fixed slapd ACL peername/sockname exact match bug (ITS#1516)
	Fixed back-passwd db_config bug
	Fixed -lldap cache debug bug (ITS#1501)
	Fixed -lldap dnssrv free bug
	Fixed -lldap Start TLS first op bug
	Fixed slurpd TLS non-critical/critical bug (ITS#1527)
	Fixed slurpd SASL password bug (ITS#1528)
	Build environment
		Fixed repl_user build error (ITS#1503)
		Updated BerkeleyDB 4 support
		Updated schema files
		Updated Windows NT/2000 MSVC5 port

OpenLDAP 2.0.19 Release
	Fixed back-ldbm not filter indexing (ITS#1405)
	Fixed back-passwd e_ndn bug (ITS#1467)
	Fixed back-ldap undef ad bug (ITS#1367)
	Fixed slapd suffix "" bugs (ITS#1430)
	Fixed slapd zero length replace bug (ITS#1364)
	Fixed slapd teletex string printables bug (ITS#1419)
	Fixed slapd unallocated sockets bug (ITS#1415)
	Fixed slapd undefined present filter bug (ITS#1439)
	Fixed -lldap cacertfile bug (ITS#1406)
	Fixed -lldap ldap_url_seach_s bug (ITS#1429)
	Fixed -lldap unbind free bug (ITS#1442)
	Fixed -llber realloc bug (ITS#1410)
	Fixed ldappasswd/slappasswd strncmp bug (ITS#1497)
	Fixed various memory leaks
	Updated slapd emfile/enfile handling
	Build environment
		Fixed passwd-search port bug (ITS#1461)
		Fixed Kerberos build bug (ITS#1468)
		Updated CThread detection (ITS#1440)
		Updated MAXPATHLEN handling (ITS#1440)
		Fixed TLS build (ITS#1412)
		Added BerkeleyDB 4 support
		Added GNU ar support (ITS#1460)

OpenLDAP 2.0.18 Release
	Fixed -llber ber_bvecadd bug (ITS#1398)
	Fixed slap tools replica config bug (ITS#1385)
	Build environment
		Updated req_query detection

OpenLDAP 2.0.17 Release
	Fixed ldbm BerkeleyDB2 compat bug (ITS#1390)

OpenLDAP 2.0.16 Release
	Fixed slapd indexing bug (ITS#1349)
	Fixed slapd/slurpd maxargs bug (ITS#1343)
	Fixed slapd ManageDSAIT warning
	Fixed slapd no OID schema bug
	Fixed slurpd server down reject bug (ITS#1183)
	Fixed slurpd moddn newsup bug (ITS#1379)
	Fixed -llber ber_realloc bug (ITS#1346)
	Fixed -lldap multi-level referral chase bug (ITS#1346)
	Fixed -lldap high port bug (ITS#1375)
	Fixed CR/LF handling (ITS#1328)
	Updated slapd/-lldap getaddrinfo handling (ITS#1304)
	Updated -lldap_r pthread stack size
	Updated slapd acl assert checks
	Updated slurpd ManageDSAIT handling
	Updated slapd root/subschema DSE update error handling
	Removed -lldbm use of Berkeley CDB
	Updated -lldbm to use r/w lock with DB_THREAD
	Updated back-ldbm update attribute indices on delete (ITS#1359)
	Updated ud Kerberos ticket handling
	Added back-ldbm idl_check() code
	Build environment
		Updated ltconfig for MacOS X 10.1 (ITS#1361)
		Updated slapd.conf with ACL example
		Updated pthread detection
		Updated Kerberos detection (ITS#1372)
		Updated test005 /dev/null diff for Tru64 compat
		Updated test007 modrdn test
		Updated test008 read/write ratio
		Updated MSVC5 project files
	Documentation
		Update slapcat/add pages

OpenLDAP 2.0.15 Release
	Fixed -lldap TLS external handling

servers/slapd/slapschema.c
/* $OpenLDAP: pkg/ldap/servers/slapd/slapschema.c,v 1.1.2.5 2010/04/14 22:59:10 quanah Exp $ */
/* This work is part of OpenLDAP Software <http://www.openldap.org/>.
 *
 * Copyright 1998-2011 The OpenLDAP Foundation.
 * Portions Copyright 1998-2003 Kurt D. Zeilenga.
 * Portions Copyright 2003 IBM Corporation.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted only as authorized by the OpenLDAP
 * Public License.
 *
 * A copy of this license is available in file LICENSE in the
 * top-level directory of the distribution or, alternatively, at
 * <http://www.OpenLDAP.org/license.html>.
 */
/* ACKNOWLEDGEMENTS:
 * This work was initially developed by Pierangelo Masarati for inclusion
 * in OpenLDAP Software.  Code portions borrowed from slapcat.c;
 * contributors are Kurt Zeilenga and Jong Hyuk Choi
 */

#include "portable.h"

#include <stdio.h>

#include "ac/stdlib.h"
#include "ac/ctype.h"
#include "ac/socket.h"
#include "ac/string.h"

#include "slapcommon.h"
#include "ldif.h"

static volatile sig_atomic_t gotsig;

static RETSIGTYPE
slapcat_sig( int sig )
{
	gotsig=1;
}

int
slapschema( int argc, char **argv )
{
	ID id;
	int rc = EXIT_SUCCESS;
	const char *progname = "slapschema";
	Connection conn = { 0 };
	OperationBuffer	opbuf;
	Operation *op = NULL;
	void *thrctx;
	int requestBSF = 0;
	int doBSF = 0;

	slap_tool_init( progname, SLAPCAT, argc, argv );

	requestBSF = ( sub_ndn.bv_len || filter );

#ifdef SIGPIPE
	(void) SIGNAL( SIGPIPE, slapcat_sig );
#endif
#ifdef SIGHUP
	(void) SIGNAL( SIGHUP, slapcat_sig );
#endif
	(void) SIGNAL( SIGINT, slapcat_sig );
	(void) SIGNAL( SIGTERM, slapcat_sig );

	if( !be->be_entry_open ||
		!be->be_entry_close ||
		!( be->be_entry_first || be->be_entry_first_x ) ||
		!be->be_entry_next ||
		!be->be_entry_get )
	{
		fprintf( stderr, "%s: database doesn't support necessary operations.\n",
			progname );
		exit( EXIT_FAILURE );
	}

	if( be->be_entry_open( be, 0 ) != 0 ) {
		fprintf( stderr, "%s: could not open database.\n",
			progname );
		exit( EXIT_FAILURE );
	}

	thrctx = ldap_pvt_thread_pool_context();
	connection_fake_init( &conn, &opbuf, thrctx );
	op = &opbuf.ob_op;
	op->o_tmpmemctx = NULL;
	op->o_bd = be;


	if ( !requestBSF && be->be_entry_first ) {
		id = be->be_entry_first( be );

	} else {
		if ( be->be_entry_first_x ) {
			id = be->be_entry_first_x( be,
				sub_ndn.bv_len ? &sub_ndn : NULL, scope, filter );

		} else {
			assert( be->be_entry_first != NULL );
			doBSF = 1;
			id = be->be_entry_first( be );
		}
	}

	for ( ; id != NOID; id = be->be_entry_next( be ) ) {
		Entry* e;
		char textbuf[SLAP_TEXT_BUFLEN];
		size_t textlen = sizeof(textbuf);
		const char *text = NULL;

		if ( gotsig )
			break;

		e = be->be_entry_get( be, id );
		if ( e == NULL ) {
			printf("# no data for entry id=%08lx\n\n", (long) id );
			rc = EXIT_FAILURE;
			if( continuemode ) continue;
			break;
		}

		if ( doBSF ) {
			if ( sub_ndn.bv_len && !dnIsSuffixScope( &e->e_nname, &sub_ndn, scope ) )
			{
				be_entry_release_r( op, e );
				continue;
			}


			if ( filter != NULL ) {
				int rc = test_filter( NULL, e, filter );
				if ( rc != LDAP_COMPARE_TRUE ) {
					be_entry_release_r( op, e );
					continue;
				}
			}
		}

		if( verbose ) {
			printf( "# id=%08lx\n", (long) id );
		}

		rc = entry_schema_check( op, e, NULL, 0, 0, NULL,
			&text, textbuf, textlen );
		if ( rc != LDAP_SUCCESS ) {
			fprintf( ldiffp->fp, "# (%d) %s%s%s\n",
				rc, ldap_err2string( rc ),
				text ? ": " : "",
				text ? text : "" );
			fprintf( ldiffp->fp, "dn: %s\n\n", e->e_name.bv_val );
		}

		be_entry_release_r( op, e );
	}

	be->be_entry_close( be );

	if ( slap_tool_destroy() )
		rc = EXIT_FAILURE;

	return rc;
}
